File: web/src/api/TransportMidi.mjs
import { toSysEx, fromSysEx, kSysExAnyChannel } from "./MidiSysEx.mjs"

export default class TransportMidi {
  constructor(navigator, connectionEvent) {
    this.navigator = navigator
    this.midiChannel = navigator.midiChannel ?? kSysExAnyChannel
    this.connectionEvent = connectionEvent
    this.midi = null
    this.input = null
    this.output = null
    this.queue = []
    this.resolve = null
    this.reject = null
  }

  get connected() {
    return this.input != null
  }

  version = _ => ({
    major: 1,
    minor: 2,
    subminor: 3,
  })

  async connect() {
    console.log('Connecting MIDI ...');
    this.midi = await this.navigator.requestMIDIAccess({sysex: true})
    const name = this.navigator.midiDeviceName
    
    console.log('inputs')
    this.midi.inputs.values().forEach(k => console.log(k))
    console.log('outputs')
    this.midi.outputs.values().forEach(k => console.log(k))

    this.input = this.midi.inputs.values().find(x => x.name === name)
    this.output = this.midi.outputs.values().find(x => x.name === name)
    if (!this.input || !this.output) {
      throw new Error(`Cannot find MIDI device '${name}'`)
    }
    this.input.onmidimessage = e => {
      console.log('Received midi message', e.data)
      const buffer = fromSysEx(e.data, this.midiChannel)
      if (!buffer) {
        console.log('Invalid sysex')
        return
      }
      const data = new DataView(buffer.buffer)
      this.queue.push(data)
      if (this.resolve) {
        this.resolve(this.queue.shift())
        this.resolve = null
        this.reject = null
      }
    }
    this.connectionEvent(true);
  }

  async reconnect() {
    return this.connect()
  }

  async send(data) {
    const buffer = new Uint8Array(data.buffer)
    const sysex = toSysEx(buffer, this.midiChannel)
    console.log('sending', Array.from(sysex))
    await this.output.send(Array.from(sysex))
  };

  async receive() {
    if (this.queue.length > 0) {
      return this.queue.shift();
    }
    return new Promise((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });
  };
}

File: tools/pytocatapedal/src/pytocatapedal/transport_midi.py
"""python-rtmidi backed MIDI transport, mirroring web/src/api/TransportNodeMidi.mjs.

python-rtmidi is the direct analog of the RtMidi-backed `midi` npm package
the Node CLI uses.
"""
import queue
from typing import Optional

import rtmidi

from .midi_sysex import ANY_CHANNEL, from_sysex, to_sysex


class TransportMidi:
    def __init__(self, device_name: str, channel: int = ANY_CHANNEL, connection_event=None):
        self.device_name = device_name
        self.channel = channel
        self.connection_event = connection_event
        self.midi_in: Optional[rtmidi.MidiIn] = None
        self.midi_out: Optional[rtmidi.MidiOut] = None
        self._queue: "queue.Queue[bytes]" = queue.Queue()

    @property
    def connected(self) -> bool:
        return self.midi_in is not None

    def version(self):
        # Hardcoded stub, matching the JS transport (never read from the device).
        return {"major": 1, "minor": 2, "subminor": 3}

    def connect(self):
        self.midi_in = rtmidi.MidiIn()
        self.midi_out = rtmidi.MidiOut()

        input_index = self._find_port(self.midi_in, self.device_name)
        output_index = self._find_port(self.midi_out, self.device_name)
        if input_index is None or output_index is None:
            raise RuntimeError(f"Cannot find MIDI device '{self.device_name}'")

        self.midi_in.set_callback(self._on_message)
        self.midi_in.open_port(input_index)
        self.midi_in.ignore_types(sysex=False, timing=True, active_sense=True)
        self.midi_out.open_port(output_index)

        if self.connection_event:
            self.connection_event(True)

    def reconnect(self):
        return self.connect()

    def send(self, data: bytes):
        sysex = to_sysex(bytes(data), self.channel)
        self.midi_out.send_message(list(sysex))

    def receive(self) -> bytes:
        return self._queue.get()

    def _on_message(self, event, _data=None):
        message, _delta_time = event
        buffer = from_sysex(bytes(message), self.channel)
        if buffer is None:
            return
        self._queue.put(buffer)

    @staticmethod
    def _find_port(port, name: str) -> Optional[int]:
        count = port.get_port_count()
        for i in range(count):
            if port.get_port_name(i) == name:
                return i
        # rtmidi backends (notably ALSA) sometimes suffix port names
        # (e.g. "Tocata Pedal:0"), so fall back to a prefix match.
        for i in range(count):
            if port.get_port_name(i).startswith(name):
                return i
        return None
